src/niMidi.cpp

```diff
@@ -234,21 +311,58 @@ class NiMidiSurface: public BaseSurface {
 			this->_sendSysex(CMD_TRACK_NAME, 0, numInBank, name);
 			// todo: level meters, volume, pan
 		}
-		// todo: navigate tracks, navigate banks
+		// todo: navigate tracks, navigate banks. NOTE: probably not here
+	}
+
+	void ClearSelected() {
+		// Clear all selected tracks. Copyright (c) 2010 and later Tim Payne (SWS)
+		int iSel = 0;
+		for (int i = 0; i <= GetNumTracks(); i++) // really ALL tracks, hence no use of CSurf_NumTracks
+			GetSetMediaTrackInfo(CSurf_TrackFromID(i, false), "I_SELECTED", &iSel);
+	}
+
+	void _onTrackSelect(unsigned char numInBank) {
+		int id = this->_bankStart + numInBank;
+		if (id > GetNumTracks()) {
+			return;
+		}
+		MediaTrack* track = CSurf_TrackFromID(id, false);
+		int iSel = 1; // "Select"
+		// If we rather wanted to "Toggle" than just "Select" we would use:
+		// int iSel = *(int*)GetSetMediaTrackInfo(track, "I_SELECTED", nullptr) ? 0 : 1; 
+		ClearSelected(); 
+		GetSetMediaTrackInfo(track, "I_SELECTED", &iSel);		
+	}
+
+	void _onBankSelect(signed char value) {
+		// Manually switch the bank visible in Mixer View WITHOUT influencing track selection
+		int newBankStart = this->_bankStart + (value * BANK_NUM_TRACKS);
+		int numTracks = CSurf_NumTracks(false); // If we ever want to show just MCP tracks in KK Mixer View (param) must be (true)
+		if ((newBankStart < 0) || (newBankStart > numTracks)) {
+			return;
+		}
+		this->_bankStart = newBankStart;
+		this->_MixerUpdate();
 	}
 
 	void _onKnobVolumeChange(unsigned char command, signed char value) {
-		int numInBank = command - CMD_KNOB_VOLUME1 + 1;
+		int numInBank = command - CMD_KNOB_VOLUME1;
+		double dvalue = static_cast<double>(value);
 		MediaTrack* track = CSurf_TrackFromID(numInBank + this->_bankStart, false);
-		if (!track) return;
-		CSurf_SetSurfaceVolume(track, CSurf_OnVolumeChange(track, value * 0.1, true), nullptr);
+		if (!track) {
+			return;
+		}
+		CSurf_OnVolumeChange(track, dvalue * 0.01, true); 
 	}
 
 	void _onKnobPanChange(unsigned char command, signed char value) {
-		int numInBank = command - CMD_KNOB_VOLUME1 + 1;
+		int numInBank = command - CMD_KNOB_PAN1;
+		double dvalue = static_cast<double>(value);
```

Comment:
Why is this casted variable needed?

Reply:
This is just a "safety belt" to avoid ambiguous interpretation by different compilers. Maybe not necessary at all (again, 23 years "without").
In the following scaling dvalue * 0.00098425 the floating point multiplier together with the function declaration probably instructs the compiler anyway to convert the signed char into a double. Just wanted to make sure it compiles correctly everywhere.